### PR TITLE
Link header

### PR DESCRIPTION
**Pull request by ioganegambaputifonguser**
Enhancemet #760  Add simple link header with URI-reference only for now. Add test with such header.

### CODE REVIEW

File: blaze-server/src/test/scala/org/http4s/server/blaze/Http1ServerStageSpec.scala [hunk 1]
```diff
@@ -5,13 +5,15 @@ import cats.effect._
 import cats.implicits._
 import java.nio.ByteBuffer
 import java.nio.charset.StandardCharsets
+
```
Comment (aeons) [inline blaze-server/src/test/scala/org/http4s/server/blaze/Http1ServerStageSpec.scala]: Imports should be sorted alphabetically without any blank lines.

**Comment by aeons** [comment]
This is a good start.

I don't know much about the Link header, but from skimming the [RFC linked to in the issue](https://tools.ietf.org/html/rfc5988#section-5), it seems the parser is a bit too simplistic, and will fail on most of the examples from the RFC.

File: blaze-server/src/test/scala/org/http4s/server/blaze/Http1ServerStageSpec.scala (hunk 1 above)
Comment (ioganegambaputifonguser) [reply blaze-server/src/test/scala/org/http4s/server/blaze/Http1ServerStageSpec.scala]: got it
btw do you know why build fails with my `Link.scala` with the format issues i guess?
locally it passed with the `test:scalafmt`

Comment (aeons) [reply blaze-server/src/test/scala/org/http4s/server/blaze/Http1ServerStageSpec.scala]: `test:scalafmt` only formats test sources. You need to also run `scalafmt` to format main sources.

You can run `sbt ci` to do the same as what ci runs. It takes a while though.

**Comment by ioganegambaputifonguser** [comment]
sure, I have to put WIP status 
Also I am going to add a few link params within the current PR, using the RFC what you mentioned

```
Link           = "Link" ":" #link-value
  link-value     = "<" URI-Reference ">" *( ";" link-param )
  link-param     = ( ( "rel" "=" relation-types )
                 | ( "anchor" "=" <"> URI-Reference <"> )
                 | ( "rev" "=" relation-types )
                 | ( "hreflang" "=" Language-Tag )
                 | ( "media" "=" ( MediaDesc | ( <"> MediaDesc <"> ) ) )
                 | ( "title" "=" quoted-string )
                 | ( "title*" "=" ext-value )
                 | ( "type" "=" ( media-type | quoted-mt ) )
                 | ( link-extension ) )
```

File: blaze-server/src/test/scala/org/http4s/server/blaze/Http1ServerStageSpec.scala [hunk 2]
```diff
@@ -413,6 +413,20 @@ class Http1ServerStageSpec extends Http4sSpec {
...
+        val service = HttpService[IO] {
+          case req => IO.pure(Response(body = req.body).replaceAllHeaders(linkHeader))
+        }
+
+        val req = "GET /foo HTTP/1.1\r\n\r\n"
+
+        val buf = Await.result(runRequest(Seq(req), service), 5.seconds)
+        val (_, hdrs, _) = ResponseParser.apply(buf)
+        hdrs.find(_.name == Link.name) must beSome(linkHeader)
+      }
```
Comment (rossabaker) [inline blaze-server/src/test/scala/org/http4s/server/blaze/Http1ServerStageSpec.scala]: This header shouldn't really need to be tested in blaze.

If you can add an `Arbitrary[Link]` (I'll help if you can't), then you can follow a model more like the [other header specs](https://github.com/http4s/http4s/blob/master/tests/src/test/scala/org/http4s/headers/ContentLengthSpec.scala).  It'll be a simple test that covers more.

**Comment by ioganegambaputifonguser** [comment]
@rossabaker somehow I missed a notification that there are some updates. So I am here and ready to finish with it. I added Arbitrary[Link] with the uri only support. But some of the tests fails. Could you please support a little with it? thanks!

**Comment by rossabaker** [comment]
Sorry, now it was my turn to miss a notification.

This failure is not obvious to me, but I strongly suspect it's related to the problem described in #1651.  If a URI doesn't round trip through the parser as the same structure, then the header containing them will also have problems.  I think if we fix that, this will also begin to work.

Yet another reason we need to fix our `Uri` type. :(

**Comment by booders-12** [comment]
_Sent from my Galaxy Tab A (2016) using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)

**Comment by rossabaker** [comment]
Wow, this one fell off the radar.  I'm so sorry.

The proper fix depends on fixing our `Uri`, but we've been busy with fs2 and cats-effect.  Is this something that we could mark the tests pending and get the header?

**Pull request merged**